src/server/cluster/incoming_slot_migration.cc

```diff
@@ -85,15 +95,28 @@ IncomingSlotMigration::IncomingSlotMigration(string source_id, Service* se, Slot
 }
 
 IncomingSlotMigration::~IncomingSlotMigration() {
-  sync_fb_.JoinIfNeeded();
 }
 
 void IncomingSlotMigration::Join() {
   bc_->Wait();
   state_ = MigrationState::C_FINISHED;
 }
 
-void IncomingSlotMigration::StartFlow(uint32_t shard, io::Source* source) {
+void IncomingSlotMigration::Cancel() {
+  LOG(INFO) << "Cancelling incoming migration of slots " << SlotRange::ToString(slots_);
+  cntx_.Cancel();
+
+  auto cb = [this](util::ProactorBase* pb) {
+    if (const auto* shard = EngineShard::tlocal(); shard) {
+      if (auto& flow = shard_flows_[shard->shard_id()]; flow) {
+        flow->Cancel();
+      }
+    }
+  };
```

Comment:
we can not use shard_id for incoming migrations, we don't know where every flow is executed

Reply:
you are right! good catch
furthermore, this isn't even needed, inside the inner Cancel() I forward to the correct proactor.